### PR TITLE
SNOW-1458130 Implement `Index.sort_values`

### PR DESCRIPTION
**Pull request by sfc-gh-vbudati**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1458130

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Added support for `Index.sort_values`. Currently, the `key` parameter is not supported in Snowpark pandas.

One peculiar feature about `Index.sort_values` is the option to return an indexer (`return_indexer`) - this is a numpy array with the old row numbers/positions in their sorted positions. 

I reused the `sort_index` implementation in the query compiler since it's the exact same logic. I added a new parameter `include_indexer` to retrieve the indexer used to sort the Index values. It is import to note here that pandas uses quicksort (which Snowpark pandas does not support yet) while Snowpark pandas uses stable sort. Therefore, it is not guaranteed that the indexer returned by Snowpark pandas is going to be the same as the one returned by native pandas.

Examples:
```py
>>> res = pd.Index([1, 3, 1, 1, 1, 3, 1, 1, 1, 2, 2, 2, 3]).sort_values(return_indexer=True)
>>> res[0]
Index([1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 3, 3, 3], dtype='int64')
>>> res[1]
array([ 0,  2,  3,  4,  6,  7,  8,  9, 10, 11,  1,  5, 12])

>>> idx = pd.Index([1, 2, 3, 2, 3, 5, 6, 7, 8, 4, 4, 5, 6, 7, 1, 2, 1, 2, 3, 4, 3, 4, 5, 6, 7])
>>> res = idx.sort_values(return_indexer=True)
# Both Snowpark pandas and native pandas return the correct Index.
>>> res[0]
Index([1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 6, 6, 6, 7, 7, 7, 8], dtype='int64')
# But the indexer order returned in this example differs since Snowpark pandas uses stable sort
# while native pandas uses quicksort.
>>> res[1]
# Snowpark pandas
array([0, 14, 16, 1, 3, 15, 17, 2, 4, 18, 20, 9, 10, 19, 21, 5, 11, 22, 6, 12, 23, 7, 13, 24, 8])
                  ^ differs
# Native pandas
array([0, 14, 16, 15, 17, 3, 1, 18, 2, 4, 20, 21, 9, 10, 19, 5, 22, 11, 12, 23, 6, 7, 13, 24, 8])
                   ^ differs
```

### CODE REVIEW

File: tests/integ/modin/index/test_sort_values.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        return_indexer=True, ascending=ascending, na_position=na_position
+    )
+
+    # Compare the index part of the results.
+    assert_index_equal(snow_res[0], native_res[0])
+
+    # Pandas performs quicksort to generate the indexer. Currently, Snowpark pandas does not support quicksort;
+    # instead, stable sort is performed. Therefore, the results are not guaranteed to match.
+    # In case the Snowpark pandas result does not match pandas, verify that it matches `np.argsort(kind=stable)`.
+    if match_pandas:
```
Comment (sfc-gh-nkumar) [inline tests/integ/modin/index/test_sort_values.py:188]: match_pandas is True for the indices without duplicates?
Given pandas output is non -stable if index has duplicates we should consider doing np.argesort(kind=stable) whenever duplicate values are present.

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 2]
```diff
@@ -1799,18 +1807,31 @@ def sort_values(
...
+            # objects to return.
+            return (
+                Index(res, convert_to_lazy=self.is_lazy),
+                Series(query_compiler=res).to_numpy(),
+            )
         else:
-            return Index(ret, convert_to_lazy=self.is_lazy)
+            # When `return_indexer` is False, a query compiler with only one index column
+            # is returned.
+            return Index(res, convert_to_lazy=self.is_lazy)
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:1834]: nit: we can deduplicate index creation 
```
index = Index(res, convert_to_lazy=self.is_lazy)
if return_indexer:
  return index, Series(query_compiler=res).to_numpy()
else:
  index
```

**Review by sfc-gh-nkumar — COMMENTED** [review]
Looks pretty clean. Thanks for adding descriptive comments.

File: tests/integ/modin/index/test_sort_values.py (hunk 1 above)
Comment (sfc-gh-vbudati) [reply tests/integ/modin/index/test_sort_values.py:188]: No, all the example tests have duplicates in them. It's just that when `match_pandas=True`, quicksort and stable sort produce the same output/indexer.

Comment (sfc-gh-nkumar) [reply tests/integ/modin/index/test_sort_values.py:188]: I see. One potential issue we have that test can become flaky when output of quicksort changes (due to non-stable nature).

Comment (sfc-gh-nkumar) [reply tests/integ/modin/index/test_sort_values.py:188]: If all have duplicate values. I think we should just use what you have in else block to compare results.

Comment (sfc-gh-vbudati) [reply tests/integ/modin/index/test_sort_values.py:188]: okay, I'll change it to compare with only numpy stable argsort

**Review by sfc-gh-nkumar — APPROVED** [review]

File: docs/source/modin/supported/index_supported.rst [hunk 3]
```diff
@@ -159,7 +159,8 @@ Methods
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
 | ``searchsorted``            | N                               |                                  |                                                    |
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
-| ``sort_values``             | N                               |                                  |                                                    |
+| ``sort_values``             | P                               | key                              | Snowpark pandas currently uses stable sort when    |
+|                             |                                 |                                  | sorting the index values. Pandas uses quicksort.   |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/supported/index_supported.rst:163]: Does this comment also apply to `Series.sort_values` and `DataFrame.sort_values`? If so, then we should update their docs too. It can be in a separate PR though.

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM. Thanks, Varnika.

File: docs/source/modin/supported/index_supported.rst (hunk 3 above)
Comment (sfc-gh-vbudati) [reply docs/source/modin/supported/index_supported.rst:163]: Yeah! Should apply there as well, I'll make a separate PR for this

Comment (sfc-gh-vbudati) [reply docs/source/modin/supported/index_supported.rst:163]: https://github.com/snowflakedb/snowpark-python/pull/1916/files
Here's the PR!
